### DISCUSSION TITLE
[Walkthrough] Flying Drills: Chase the Rabbit

### DISCUSSION THREAD

**Discussion by cbot-coder** [colobot/colobot#1379] [Workshop]
## Overview

The Flying Drills (Moon level 2 and Moon level 3) are almost identical. At first, I really struggled with how to make the bot fly smoothly through the targets. But finally I hit upon an approach that works brilliantly.

The basic idea is this: To get the bot to steer a bit to the side of the target loop as it approaches, then turn and swoop through smoothly, I imagined a spot on a line perpendicular to the target, but some distance in front of it. The bot would aim for it first. In the illustration below, this is point A (not to scale).
![image](https://user-images.githubusercontent.com/75964590/104361007-802bfb80-54d7-11eb-8315-656cddd18ed7.png)

Then, as the bot gets closer to the target, the spot moves along that perpendicular line closer to the target as well (point B, then C), eventually passing through the target to the other side (point D and so on). The bot continues to "chase" that point, **like a dog chasing a rabbit as the rabbit darts through a hole in the fence**.

That takes care of the X, Y steering of the bot. For the height, I simply had it aim to match the height of the target loop as it approached.

I was very pleased with how smoothly and accurately this approach works.

## Code

Here is the solution in full:

```c++
extern void object::SolveFlyingDrill() {

  // This program flies through the Flying Drill #1 and #2 (Moon
  // levels 2 and 3).
  //
  // It doesn't need to stop, and doesn't bump into anything.
  // It is very smooth and precise.
  // Most of the numbers were selected empirically.

  object target;
  float dist, dir, trueorient, rabbitoffset, height, maxspeed = 1, speed;
  point rabbit;

  while ((target = radar(Target2, 0, 180)) != null) {

    // Get distance and height of target
    dist = distance2d(position, target.position);
    height = target.position.z - 1; // compensate for height of bot

    // To fly a smooth curve going through the target, passing through
    // the loop as perpendicular as possible, start by aiming for a
    // spot slightly in front of the target. Then, have that spot move
    // *through* the target to the other side as we approach. Think of
    // it as the bot chasing a rabbit who is going through the hoop.
    rabbitoffset = (dist > 25) ? 10 : dist - 15;
    trueorient = target.orientation - 90;
    rabbit.x = target.position.x + cos(trueorient) * rabbitoffset;
    rabbit.y = target.position.y + sin(trueorient) * rabbitoffset;

    // Sometimes the Target2's orientation is 180 degrees wrong. Check
    // and correct for that here:
    if ((rabbitoffset > 0 and distance2d(position, rabbit) > dist) or
        (rabbitoffset < 0 and distance2d(position, rabbit) < dist)) {
      trueorient += 180;
      rabbit.x = target.position.x + cos(trueorient) * rabbitoffset;
      rabbit.y = target.position.y + sin(trueorient) * rabbitoffset;
    }

    // Uncomment this line to visualize where it's aiming
    //produce(rabbit, 0, WayPoint, "test");

    // Get the direction toward the imaginary rabbit's position
    dir = direction(rabbit);

    // Keep approaching the height of the target itself
    // However, do it more aggressively closer to target
    jet((height - position.z) / (2 + dist / 2));

    // Slow down slightly as we approach the target
    // (Not needed anymore, as the flight is so smooth)
    //speed = (dist > 5) ? maxspeed : dist / 10 + 0.5;
    speed = maxspeed;

    // Move forward and turn
    motor(speed - (dir / 90), speed + (dir / 90));

    wait(0.1); // only need to adjust 10 times a second
  }

  // Got all the targets.  Go to ship
  target = radar(SpaceShip);
  goto(target.position);
}
```

## What about the Shooting Drill (Tropica 4)?

This flight pattern works also for Tropica 4 (Shooting Drill 1); it just doesn't yet know how to shoot the TargetBots as it flies. I leave that as an exercise for the reader!

**Comment by cbot-coder** [comment]
I found @SmileSick's solution at https://github.com/SmileSick/Colobot-Walkthrough/blob/main/2)%20Moon%20-%20Mission%202%263 and it is also fairly accurate. It doesn't bother trying to swoop smoothly through the targets, but it turns out that isn't really necessary to finish the level. :+1:

**Reply by SmileSick** [reply]
Thanks for mentioning my code.
